src/shared/Microsoft.Git.CredentialManager/HttpClientFactory.cs

```diff
@@ -85,6 +87,35 @@ public HttpClient CreateClient()
 #endif
             }
 
+            string sslCaInfo = _settings.SslCaInfo;
+            if (!string.IsNullOrWhiteSpace(sslCaInfo))
+            {
+                Func<X509Chain, SslPolicyErrors, bool> validationCallback = (chain, errors) =>
+                {
+                    if ((errors & ~SslPolicyErrors.RemoteCertificateChainErrors) != 0 ||
+                        chain.ChainStatus.Any(status => status.Status != X509ChainStatusFlags.UntrustedRoot))
+                    {
+                        // Fail validation if remote cert has any errors other than untrusted root
+                        return false;
+                    }
+
+                    // Get the custom CA root from http.sslCAInfo
+                    X509Certificate2 caBundle = new X509Certificate2(sslCaInfo);
+
+                    // Read the root of the remote certificate chain
+                    X509Certificate2 remoteRoot = chain.ChainElements[chain.ChainElements.Count - 1].Certificate;
```

Comment:
Do we want to check only the root certificate in the chain matches the user supplied one, or should we be happy that the provided certificate is _anywhere_ in the chain?

For example if the user provides cert `X` as the `sslCAInfo`, and the chain we're validating looks like this:

```text
A -> B -> X -> C -> D
```

..should this chain be "ok"?

Reply:
After looking into this, I actually got a lot more clarification on what exactly OpenSSL does with this setting ([link to the documentation](https://www.openssl.org/docs/man1.1.1/man3/SSL_CTX_set_default_verify_paths.html#NOTES)):

- Like you suggested, it should allow _any_ of the certs in the CA file, _anywhere_ in the certificate chain
- The specified file can actually contain multiple certificates (so it should probably be an `X509Certificate2Collection`)

Based on that (relating to my earlier question on chain errors) I think the way the certificate is currently verified isn't quite correct. My thought now is:

- Add the set of certificates at `sslCAInfo` to the callback `chain`'s "[ExtraStore](https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.x509certificates.x509chainpolicy.extrastore?view=net-5.0)"
- Rebuild the chain with "[Build](https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.x509certificates.x509chain.build?view=net-5.0)"
- Return the result

It looks like this might still fail with "Untrusted Root" (if the root of the chain is in the file at `sslCAInfo`) - [link](https://github.com/dotnet/runtime/issues/20302) to a relevant discussion, where a fix was unfortunately only just implemented in .NET 5.0. Without that update, I think we'd need to do approximately what I have now, which is to ignore "Untrusted Root" errors (as long as the root certificate matches one in `sslCAInfo`). That'll end up needing some redundant computation, but I should be able to mimic what OpenSSL is doing (check thumbprints of certificates in `sslCAInfo` until a match is found, then compare raw data).

Sorry for the brain dump - just want to get all of this written down so I don't forget it 🙂 Also, let me know if I missed something or if you can think of a better way to handle these things.